File: NeetCode150/Binary Search/875. Koko Eating Bananas/sol.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int H) {
        int left = 1;
        int right = 0;
        for (int i=0; i<piles.size(); i++) {
            if (piles[i] > right) {
                right = piles[i];
            }
        }        
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            int hours = 0;
            for (int pile : piles) {
                hours += (pile + (mid - 1)) / mid; // 向上取整，計算吃完一堆香蕉需要的時間
            }
            if (hours <= H) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        
        return left;
    }
};

File: NeetCode150/Binary Search/875. Koko Eating Bananas/sol.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)
        while l < r:
            k = (r - l) // 2 + l
            hours = 0
            for pile in piles:
                hours += (pile - 1) // k + 1
            if hours > h:
                l = k + 1
            else:
                r = k
        return l      